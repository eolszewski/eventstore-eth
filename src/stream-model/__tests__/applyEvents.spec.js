const { EventStore } = require('../../index');
const config = require('../../config');
const eventStoreArtifact = require('../../../build/contracts/EventStore.json');

const eventStore = new EventStore({
  eventStoreArtifact,
  ...config
});

const StreamModel = require('../index');

const events = require('../../__mock__/events.json');

describe('applyEvents', () => {
  let accounts;
  let eventStore;
  let streamModel;

  const filter = event => {
    return true;
  };
  const reducer = (state, event) => {
    return {
      ...state,
      events: [event.value, ...(state.events || [])],
      eventsProcessed: (state.eventsProcessed || 0) + 1
    };
  };

  beforeAll(async () => {
    eventStore = new EventStore({
      eventStoreArtifact,
      ...config
    });
    await eventStore.init();
    accounts = await eventStore.getWeb3Accounts();
  });

  // new event store per test
  beforeEach(async () => {
    eventStore = await eventStore.clone(accounts[0]);
    streamModel = new StreamModel(eventStore, filter, reducer);
  });

  it('should be idempotent', async () => {
    streamModel.applyEvents(events);
    expect(streamModel.state.model).toMatchSnapshot();
    streamModel.applyEvents(events);
    expect(streamModel.state.model).toMatchSnapshot();
  });

  it('should support custom event filters', async () => {
    const custom_filter = event => {
      return event.key.type === 'patient' && event.key.id === '0';
    };
    const streamModel = new StreamModel(eventStore, custom_filter, reducer);
    streamModel.applyEvents(events);
    expect(streamModel.state.model).toMatchSnapshot();
  });
});
