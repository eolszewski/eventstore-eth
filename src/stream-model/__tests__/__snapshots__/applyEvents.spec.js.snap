// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applyEvents should be idempotent 1`] = `
Object {
  "events": Array [
    Object {
      "type": "USER_UPDATED",
      "username": "bob2@example.com",
    },
    Object {
      "guid": "0x0...",
      "type": "DEVICE_LINKED",
    },
    Object {
      "type": "USER_REGISTERED",
      "username": "alice@example.com",
    },
    Object {
      "type": "USER_REGISTERED",
      "username": "bob@example.com",
    },
    Object {
      "guid": "0x0...",
      "type": "DEVICE_REGISTERED",
    },
  ],
  "eventsProcessed": 5,
}
`;

exports[`applyEvents should be idempotent 2`] = `
Object {
  "events": Array [
    Object {
      "type": "USER_UPDATED",
      "username": "bob2@example.com",
    },
    Object {
      "guid": "0x0...",
      "type": "DEVICE_LINKED",
    },
    Object {
      "type": "USER_REGISTERED",
      "username": "alice@example.com",
    },
    Object {
      "type": "USER_REGISTERED",
      "username": "bob@example.com",
    },
    Object {
      "guid": "0x0...",
      "type": "DEVICE_REGISTERED",
    },
    Object {
      "type": "USER_UPDATED",
      "username": "bob2@example.com",
    },
    Object {
      "guid": "0x0...",
      "type": "DEVICE_LINKED",
    },
    Object {
      "type": "USER_REGISTERED",
      "username": "alice@example.com",
    },
    Object {
      "type": "USER_REGISTERED",
      "username": "bob@example.com",
    },
    Object {
      "guid": "0x0...",
      "type": "DEVICE_REGISTERED",
    },
  ],
  "eventsProcessed": 10,
}
`;

exports[`applyEvents should support custom event filters 1`] = `
Object {
  "events": Array [
    Object {
      "type": "USER_UPDATED",
      "username": "bob2@example.com",
    },
    Object {
      "guid": "0x0...",
      "type": "DEVICE_LINKED",
    },
    Object {
      "type": "USER_REGISTERED",
      "username": "bob@example.com",
    },
  ],
  "eventsProcessed": 3,
}
`;
